File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.java
class Solution {
    public long countFairPairs(int[] nums, int lower, int upper) {
        //so we are sorting, Why??
        //because nums[i] + nums[j] == nums[j] + nums[i]
        //therefore, it doesn't matter
        Arrays.sort(nums);

        int lowerPairs = 0, upperPairs = 0;
        // int right = nums.length;
        long res = 0;
        int n = nums.length;

        for(int i = 0; i < nums.length; i++){
            lowerPairs = lowerBound(nums, i+1, n, lower - nums[i]);
            upperPairs = upperBound(nums, i+1, n, upper - nums[i]);
            res += upperPairs - lowerPairs;
        }
        return res;
    }

    private int lowerBound(int[] nums, int left, int right, int target){
        // int res = 0;
        while(left < right){
            int mid = left+ (right-left)/2;

            if(nums[mid] >= target){
                // res = mid;
                right = mid;
            }else{
                left = mid + 1;
            }
        }
        return left;
    }
    private int upperBound(int[] nums, int left, int right, int target){
       
        while(left < right){
            int mid = left+ (right-left)/2;

            if(nums[mid] > target){
                right = mid;
            }else{
                left = mid + 1;
            }
        }
        return left;
    }
}

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.rs
impl Solution {
    pub fn count_fair_pairs(nums: Vec<i32>, lower: i32, upper: i32) -> i64 {
        let mut nums = nums;
        let n = nums.len();
        let mut lowerPairs = 0;
        let mut upperPairs = 0;
        nums.sort();
        let mut count : i64 = 0;

        for i in 0..n{
            lowerPairs = Self::lower_bound(&nums, i+1, n, lower - nums[i]);
            upperPairs= Self::upper_bound(&nums, i+1, n, upper - nums[i]);
            count += upperPairs - lowerPairs;
        }
        count
    }

    fn lower_bound(nums: &Vec<i32>, mut left : usize, mut right : usize, target: i32) -> i64{
        while left < right{
            let mid = left + (right - left)/2;
            if nums[mid] >= target{
                right = mid;
            }else{
                left = mid + 1;
            }
        }
        left as i64
    }
    fn upper_bound(nums: &Vec<i32>, mut left : usize, mut right : usize, target: i32) -> i64{
        while left < right{
            let mid = left + (right - left)/2;
            if nums[mid] > target{
                right = mid;
            }else{
                left = mid + 1;
            }
        }
        left as i64
    }

}